### ISSUE TITLE
Support Kotlin 2.0.20

### ISSUE THREAD

**Issue by agrosner** [detekt/detekt#7679]
## Expected Behavior

Kotlin 2.0.20 is not allowed to run with detekt. it comes with a bunch of improvements that we'd like to make use of. https://blog.jetbrains.com/kotlin/2024/08/kotlin-2-0-20-released/

Is there any workaround, or must we hold off our entire project to upgrade to 2.0.20 while we wait for detekt to update?

**Comment by 3flex** [comment]
> Kotlin 2.0.20 is not allowed to run with detekt

What do you mean exactly? Even though detekt is currently using 2.0.10, you should be able to analyse your code.

**Comment by 3flex** [comment]
Also just an FYI but there's a regression in 2.0.20 so we are waiting for 2.0.21 before updating. Follow #7580 to see when it's merged and becomes available in a snapshot version.

**Comment by BraisGabin** [comment]
> Kotlin 2.0.20 is not allowed to run with detekt.

That's not true. Detekt works with kotlin 2.0.20. What problem do you have? Which is the error message?

**Comment by agrosner** [comment]
```
Execution failed for task ':&&&&&:detekt'.
> detekt was compiled with Kotlin 2.0.10 but is currently running with 2.0.20.
  This is not supported. See https://detekt.dev/docs/gettingstarted/gradle#dependencies for more information.
```

**Comment by 3flex** [comment]
And did you use the provided link which provides more information about this?

**Comment by agrosner** [comment]
Absolutely. It mentioned something about spring and `This can happen when build scripts use things like this which override all dependency configurations` . 

The doc does not make sense to me , since in this situation I’m setting the project level kotlin version. The solutions provided suggest overriding the Kotlin version for all configurations, that is not expected. Rather I’d expect a solution that allows me to run only detekt in supported kotlin version maybe?

**Comment by 3flex** [comment]
Can you please provide a build scan?

> Rather I’d expect a solution that allows me to run only detekt in supported kotlin version maybe?

This is possible as long as nothing in your Gradle config is overriding the Kotlin version in the Gradle dependency configuration that's created & used by detekt. Spring's dependency plugin overrides it, but there are other causes. This isn't a detekt problem, it's a build config problem (a build scan would confirm this).

> I’m setting the project level kotlin version

Using what method?

**Comment by 3flex** [comment]
And if you cannot provide a build scan please provide a reproducer or a suggestion/PR on how to improve our documentation.

**Comment by agrosner** [comment]
im applying kotlin plugin to the project's buildscript in a normal way via `classpath` and `apply plugin` (kotlin-android / kotlin depending on android or jvm library). 


```groovy
// top level build.gradle
buildscript {
  // ..

  dependencies {
    classpath deps.gradlePlugins.detekt
 }
}

```

in a convention plugin, that modules apply simply:
```groovy
plugins {
  id `'kotlin-android`
}
```

> This is possible as long as nothing in your Gradle config is overriding the Kotlin version in the Gradle dependency configuration that's created & used by detekt

this is an example of our detekt plugin convention script:

```

buildscript {
    repositories {
        maven {
            url "https://plugins.gradle.org/m2/"
        }
    }

    dependencies {
        classpath deps.gradlePlugins.detekt
    }
}

apply plugin: 'io.gitlab.arturbosch.detekt'

dependencies {
    detektPlugins deps.detektFormatting
}

detekt {
    toolVersion = deps.versions.detekt.get()

    config = files(configFile('quality/detekt/detekt-config.yml'))
    baseline = file('detekt-baseline.xml')
}

// based on https://detekt.github.io/detekt/gradle.html#reports
tasks.named("detekt").configure {
    reports {
        xml {
            enabled = true
            destination = file("build/reports/detekt/detekt.xml")
        }
        html {
            enabled = true
            destination = file("build/reports/detekt/detekt.html")
        }
        txt.enabled = false
    }
}

tasks.withType(io.gitlab.arturbosch.detekt.Detekt).configureEach {
    source("src/main/kotlin", "src/debug/kotlin", "src/main/java", "src/debug/java")
}

tasks.withType(io.gitlab.arturbosch.detekt.DetektCreateBaselineTask).configureEach {
    source("src/main/kotlin", "src/debug/kotlin", "src/main/java", "src/debug/java")
}
```

as you can see, we're not doing anything magical. we're just applying kotlin plugins in each of our modules with the project level kotlin version, applied via `classpath` version from a TOML definition. 

My original question was, is it possible to workaround thiso r override the kotlin version used _only_ for detekt to run with a supported version of kotlin, while maintaining newer kotlin versions. The example in the documentation makes it seem like we should only tie our project kotlin version to one that detekt supports:

```groovy
configurations.all {
    resolutionStrategy.eachDependency {
        if (requested.group == "org.jetbrains.kotlin") {
            useVersion(io.gitlab.arturbosch.detekt.getSupportedKotlinVersion())
        }
    }
}
```

and then this : 

> If Kotlin dependencies are being aligned like this then exclude the detekt dependency configuration with something like >
> configurations.matching { it.name != "detekt" }.all. 

We aren't applying the kotlin version this way.

**Comment by koriit-kontakt** [comment]
We use the following plugins:
```
plugins {
    val kotlinVersion = "2.0.20"
    kotlin("jvm") version kotlinVersion
    kotlin("plugin.spring") version kotlinVersion
    id("io.gitlab.arturbosch.detekt") version "1.23.7"
    // ...
}
```
Unfortunately, we face the same issue, and it's unclear (even after reading the provided documentation) how we can resolve it. As far as I am aware, this is a Gradle plugin problem, but to be honest, since that's the only way we actually use detekt, to us, it's one and the same.

**Comment by BraisGabin** [comment]
Just quoting 3flex comments:

> Can you please provide a build scan?

> And if you cannot provide a build scan please provide a reproducer or a suggestion/PR on how to improve our documentation.

The problem is that the spring plugin is changing the detekt dependencies (they have an issue open about this: https://github.com/spring-gradle-plugins/dependency-management-plugin/issues/369, you can leave a thumbs up there). What you should do is undo those changes. Maybe this issue helps #6198.

**Comment by koriit-kontakt** [comment]
@BraisGabin thank you for redirecting me to #6198, this workaround finally worked: https://github.com/detekt/detekt/issues/6198#issuecomment-2265183695

My problem was specifically caused by spring plugin management. The provided documentation tries to be generic but I actually needed this point solution.

**Comment by 3flex** [comment]
We've updated Kotlin to 2.0.21 in #7580.

This will be available in the next release. You can use snapshots in the meantime.

**Comment by agrosner** [comment]
thank you. any word on when a stable (non-snapshot) version with support for 2.0.21 will get released? or will we skip right to 2.1.0

**Comment by cortinico** [comment]
> thank you. any word on when a stable (non-snapshot) version with support for 2.0.21 will get released?

Nope I think we should release a version with Kotlin 2.0.21 as it will make it easier for folks to update to that version
